lib/tbd/psi.rb

```diff
@@ -1909,39 +2091,44 @@ def process(model = nil, argh = {})
           next if sh[:val].empty?
 
           edges.values.each do |edge|
-            next     if edge.key?(:io_set)
+            match = false
             next unless edge.key?(:psi)
             next unless edge.key?(:surfaces)
+            next     if edge.key?(:io_set)
 
             edge[:surfaces].keys.each do |id|
+              break    if match
               next unless tbd[:surfaces].key?(id)
               next unless tbd[:surfaces][id].key?(key)
-              next unless group[:id] == tbd[:surfaces][id][key].nameString
 
-              edge[groups]              = {} unless edge.key?(groups)
-              edge[groups][group[:psi]] = {}
-              set                       = {}
+              match = group[:id] == tbd[:surfaces][id][key].nameString
+            end
 
-              if edge.key?(:io_type)
-                safer = json[:psi].safe(group[:psi], edge[:io_type])
-                set[edge[:io_type]] = sh[:val][safer] if safer
-              else
-                edge[:psi].keys.each do |type|
-                  safer = json[:psi].safe(group[:psi], type)
-                  set[type] = sh[:val][safer]         if safer
-                end
-              end
+            next unless match
 
-              edge[groups][group[:psi]] = set unless set.empty?
+            set                       = {}
+            edge[groups]              = {} unless edge.key?(groups)
+            edge[groups][group[:psi]] = {}
+
+            if edge.key?(:io_type)
+              safer = json[:psi].safe(group[:psi], edge[:io_type])
+              set[edge[:io_type]] = sh[:val][safer] if safer
+            else
+              edge[:psi].keys.each do |type|
+                safer = json[:psi].safe(group[:psi], type)
+                set[type] = sh[:val][safer] if safer
+              end
```

Comment:
Reworking some of the previous code to skip redundant loops. Needs testing nonetheless. TO DO.